File: mcp_server/memory/migrations.py
"""One-time startup migrations, run by server.py before serving requests.

Each migration is idempotent and cheap on an already-migrated store: it
scans, finds nothing to fix, and returns. They live here rather than in
server.py so they can be unit-tested against the in-memory fakes without
importing the FastMCP app.
"""

import logging

logger = logging.getLogger(__name__)

# Articles are labelled with this project unless the feed sets its own
# `project:` in feeds.yml. Must match _DEFAULT_PROJECT in rss_worker/ingester.py.
RSS_PROJECT_LABEL = "RSS"


def migrate_missing_state(store) -> None:
    """Backfill state=active on memories that pre-date the state field.

    Recall pushes a state tag filter into FT.SEARCH; a doc with no state
    field would silently drop out of every filtered search even though the
    Python-side default treats missing state as active. One-time backfill
    keeps the two behaviours identical.
    """
    fixed = 0
    for ns in ("episodic", "project", "knowledge", "preference"):
        keys = store.scan_prefix(f"mem:{ns}:")
        if not keys:
            continue
        rows = store.get_fields_multi(keys, ("state",))
        for key, row in zip(keys, rows):
            if row is None or not row.get("state"):
                store.set_field(key, "state", "active")
                fixed += 1
    if fixed:
        logger.info("Migration: backfilled state=active on %d memories", fixed)


def migrate_project_names(store) -> None:
    """Set project_name from project field on ULID-keyed project memories missing it."""
    keys = store.scan_prefix("mem:project:")
    if not keys:
        return

    all_data = store.get_multi(keys)
    fixed = 0
    for key, data in zip(keys, all_data):
        if not data:
            continue
        # Skip entries that already have project_name set
        if data.get("project_name"):
            continue
        # Use the project field if available
        project = data.get("project")
        if project:
            store.set_field(key, "project_name", project)
            fixed += 1

    if fixed:
        logger.info("Migration: set project_name on %d project memories", fixed)


def migrate_rss_article_projects(store) -> None:
    """Label pre-existing RSS articles with project="RSS".

    The ingester stamps a project on every article (default "RSS", per-feed
    override in feeds.yml) so ingested articles can be separated from
    conversation-sourced knowledge. Articles are identified by their
    feed_name field — extracted facts and remember() writes never carry one.
    Articles that already have a project (a per-feed label) are left alone.
    """
    keys = store.scan_prefix("mem:knowledge:")
    if not keys:
        return

    rows = store.get_fields_multi(keys, ("feed_name", "project"))
    fixed = 0
    for key, row in zip(keys, rows):
        if not row:
            continue
        if row.get("feed_name") and not row.get("project"):
            store.set_field(key, "project", RSS_PROJECT_LABEL)
            fixed += 1

    if fixed:
        logger.info(
            "Migration: labelled %d RSS articles with project=%s",
            fixed, RSS_PROJECT_LABEL,
        )


def migrate_project_domains(store) -> None:
    """Seed work-type domains (v6.6) from each project's existing stack string.

    Without this, every project upgrading to v6.6 has no domains, so
    recall(domain_filter=...) matches nothing on day one — the failure mode
    that made issue #20 so hard to spot, where a filter silently returns
    nothing rather than reporting that it has nothing to filter on.

    Deliberately conservative and never destructive: it only writes to
    projects with no `domains` field at all, and it only reads a field a human
    already wrote by hand. A project whose stack yields nothing usable gets an
    empty marker so a later run doesn't rescan it, and so the web UI can tell
    "not set up yet" from "considered and empty". Anything derived here is
    editable in the web UI or replaceable with compile_project_domains().
    """
    from .project_domains import (
        invalidate_domain_cache,
        normalise_domains,
        parse_domains,
        serialise_domains,
        _STACK_STOPWORDS,
    )

    keys = store.scan_prefix("mem:project:")
    if not keys:
        return

    rows = store.get_fields_multi(keys, ("stack", "domains", "goals"))
    seeded = 0
    marked = 0
    for key, row in zip(keys, rows):
        if not row:
            continue
        # `domains` present at all (even empty) means this project has been
        # through the migration or been edited since.
        if row.get("domains") is not None:
            continue
        # Only real context entries carry a stack or goals; ULID-keyed project
        # memories are not projects and must not grow a domains field.
        if not (row.get("stack") or row.get("goals")):
            continue

        candidates = [
            item for item in parse_domains(row.get("stack") or "")
            if item.strip().lower() not in _STACK_STOPWORDS
        ]
        domains, _, _ = normalise_domains(candidates)
        store.set_field(key, "domains", serialise_domains(domains))
        if domains:
            seeded += 1
        else:
            marked += 1

    if seeded or marked:
        invalidate_domain_cache()
        logger.info(
            "Migration: seeded domains on %d projects from their stack "
            "(%d had nothing usable to derive)",
            seeded, marked,
        )


def migrate_licence(store) -> None:
    """Backfill the licence / redistribution field (v6.6.1) with honest defaults.

    Conversation-sourced namespaces (episodic, project, preference) are our
    own work and get ``own``. Knowledge is the namespace that holds
    third-party material, and it gets no optimistic guess: RSS articles
    (identified by ``feed_name``) become ``unknown`` so recall can point
    them out for a human to classify, and so do plain knowledge writes with
    no traceable origin. The one derivation the migration does make is for
    extracted facts — a fact carries ``enriched_from`` pointing at the
    memory it was extracted from, and a derivative has exactly its source's
    rights, wherever the fact landed (knowledge or preference). Sources are
    resolved from this pass first and the store second; a fact whose source
    is gone is own, because the only thing that is ever enriched is a
    conversation write.

    Memories that arrived in a skill bundle (``imported_at``) are someone
    else's work and become ``unknown``, whatever namespace they sit in.

    Idempotent: only records with no ``licence`` field at all are touched,
    so a value a human has since set is never revisited. Writes are batched
    per value through set_fields_multi. Runs at startup and again after a
    backup restore, since a pre-6.6.1 dump carries no licence field.
    """
    from .licence import LICENCE_OWN, LICENCE_UNKNOWN

    own_keys: list[str] = []
    imported_keys: list[str] = []
    pending_facts: list[tuple[str, str]] = []  # (fact key, source key)
    for ns in ("episodic", "project", "preference"):
        keys = store.scan_prefix(f"mem:{ns}:")
        if not keys:
            continue
        rows = store.get_fields_multi(keys, ("licence", "imported_at", "enriched_from"))
        for key, row in zip(keys, rows):
            row = row or {}
            if row.get("licence"):
                continue
            # A memory that arrived in a skill bundle is someone else's work
            # by definition; "own" would be exactly the optimistic guess the
            # backfill exists to avoid.
            if row.get("imported_at"):
                imported_keys.append(key)
            elif row.get("enriched_from"):
                # An extracted preference inherits like an extracted fact.
                pending_facts.append((key, row["enriched_from"]))
            else:
                own_keys.append(key)
    # Stamp the conversation namespaces first: extracted facts inherit from
    # them, so the sources must carry a value before the facts are read.
    if own_keys:
        store.set_fields_multi(own_keys, {"licence": LICENCE_OWN})
    if imported_keys:
        store.set_fields_multi(imported_keys, {"licence": LICENCE_UNKNOWN})

    knowledge_keys = store.scan_prefix("mem:knowledge:")
    knowledge_by_value: dict[str, list[str]] = {}
    rows = store.get_fields_multi(
        knowledge_keys, ("licence", "feed_name", "enriched_from", "imported_at"),
    )
    for key, row in zip(knowledge_keys, rows):
        row = row or {}
        if row.get("licence"):
            continue
        source = row.get("enriched_from")
        if source and not row.get("feed_name") and not row.get("imported_at"):
            pending_facts.append((key, source))
        else:
            # Articles, plain writes, and anything imported — including an
            # imported fact, whose source lives on another instance.
            knowledge_by_value.setdefault(LICENCE_UNKNOWN, []).append(key)

    inherited = 0
    if pending_facts:
        # Sources stamped in this pass resolve in memory; the rest with one
        # lookup per distinct source, not per fact.
        stamped = {k: LICENCE_OWN for k in own_keys}
        stamped.update({k: LICENCE_UNKNOWN for k in imported_keys})
        unresolved = sorted({src for _, src in pending_facts if src not in stamped})
        source_rows = store.get_fields_multi(unresolved, ("licence",))
        stamped.update({
            src: (row or {}).get("licence") for src, row in zip(unresolved, source_rows)
        })
        for key, source in pending_facts:
            value = stamped.get(source)
            if value:
                inherited += 1
            # A fact's only possible source is a conversation write, so a
            # source that is gone was own — the same answer the read-time
            # fallback gives, so nothing flips when the backfill runs.
            knowledge_by_value.setdefault(value or LICENCE_OWN, []).append(key)

    for value, keys in knowledge_by_value.items():
        store.set_fields_multi(keys, {"licence": value})

    unknown = len(knowledge_by_value.get(LICENCE_UNKNOWN, [])) + len(imported_keys)
    total = len(own_keys) + len(imported_keys) + sum(
        len(keys) for keys in knowledge_by_value.values()
    )
    if total:
        logger.info(
            "Migration: backfilled licence on %d memories (%d own, %d unknown, "
            "%d extracted facts inherited their source's licence)",
            total, len(own_keys), unknown, inherited,
        )


def _is_project_context(key: str, row: dict) -> bool:
    """A context entry lives at mem:project:{project_name}; a ULID-keyed
    project memory from remember(namespace="project") sets project_name
    too but never matches its own key. Stack or goals is the fallback for
    contexts written before project_name was stamped consistently."""
    name = row.get("project_name")
    if name and key == f"mem:project:{name}":
        return True
    return bool(row.get("stack") or row.get("goals"))


def migrate_provenance(store) -> None:
    """Backfill the provenance class (v6.6.2).

    Knowledge with a ``feed_name`` is an article — retrieved. Preferences
    and project context entries (keyed by their project_name) are what the
    human told us — asserted. Extracted facts (``enriched_from``, in the
    knowledge or preference namespace) are restatements and inherit their
    source, exactly as live enrichment stamps them; a fact whose source is
    gone is concluded, since it exists only because the system produced
    it. Every other record — episodic memories, ULID-keyed project
    memories, plain knowledge writes — is the agent's own account and
    becomes concluded (a plain knowledge write made today defaults to
    retrieved, but a legacy one carries no evidence of where it came from).

    That last default is the deliberate, uncomfortable choice: an episodic
    memory is the write-up of work that really happened, and calling the
    lot "concluded" undersells the most valuable material in the store.
    But it is honest about who wrote it, and the human can vouch for any
    memory afterwards with set_provenance(..., "asserted"). Idempotent:
    only records with no ``provenance`` field are touched. Runs at startup
    and after a backup restore.
    """
    from .provenance import (
        PROVENANCE_ASSERTED,
        PROVENANCE_CONCLUDED,
        PROVENANCE_RETRIEVED,
    )

    assigned: dict[str, str] = {}
    pending_facts: list[tuple[str, str]] = []  # (fact key, source key)

    # Everything that is not an extracted fact gets its class from what it
    # is. Facts are deferred: they inherit from their source, which may be
    # assigned in this same pass — so sources are resolved from the
    # in-memory assignments first and the store second, never relying on
    # the order the namespaces happened to be written in.
    for ns in ("episodic", "project", "preference", "knowledge"):
        keys = store.scan_prefix(f"mem:{ns}:")
        if not keys:
            continue
        rows = store.get_fields_multi(
            keys, ("provenance", "project_name", "stack", "goals", "feed_name", "enriched_from"),
        )
        for key, row in zip(keys, rows):
            row = row or {}
            if row.get("provenance"):
                continue
            if row.get("feed_name"):
                assigned[key] = PROVENANCE_RETRIEVED
            elif row.get("enriched_from"):
                pending_facts.append((key, row["enriched_from"]))
            elif ns == "preference" or (ns == "project" and _is_project_context(key, row)):
                assigned[key] = PROVENANCE_ASSERTED
            else:
                assigned[key] = PROVENANCE_CONCLUDED

    inherited = 0
    if pending_facts:
        # One lookup per distinct source, not per fact.
        unresolved = sorted({src for _, src in pending_facts if src not in assigned})
        source_rows = store.get_fields_multi(unresolved, ("provenance",))
        stored = {
            src: (row or {}).get("provenance") for src, row in zip(unresolved, source_rows)
        }
        for key, source in pending_facts:
            value = assigned.get(source) or stored.get(source)
            if value:
                inherited += 1
            assigned[key] = value or PROVENANCE_CONCLUDED

    by_value: dict[str, list[str]] = {}
    for key, value in assigned.items():
        by_value.setdefault(value, []).append(key)
    for value, keys in by_value.items():
        store.set_fields_multi(keys, {"provenance": value})

    total = len(assigned)
    if total:
        logger.info(
            "Migration: backfilled provenance on %d memories "
            "(%d extracted facts inherited their source's class)",
            total, inherited,
        )

File: crates/omnimem-store/src/migrations.rs
//! The 6.x startup migrations, run after an import.
//!
//! Ported from `mcp_server/memory/migrations.py`, each as a pure function over
//! every memory's fields so the rules can be tested directly, then persisted
//! in one transaction. All are idempotent: they only fill in what is missing,
//! and never bump `updated_at` (the skill compiler reads that as "source
//! changed").
//!
//! Not here yet: `migrate_project_domains`, which needs the domain
//! normalisation the engine owns (phase 3).

use std::collections::{BTreeMap, HashMap};

use omnimem_core::classification::{
    DEFAULT_CLASSIFICATION, LICENCE_OWN, LICENCE_UNKNOWN, PROVENANCE_ASSERTED,
    PROVENANCE_CONCLUDED, PROVENANCE_RETRIEVED, RSS_PROJECT_LABEL,
};
use omnimem_core::content_hash;
use serde::Serialize;
use tracing::info;

use crate::store::{Fields, memory_namespace, merge_memory, parse_fields};
use crate::{Result, Store};

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize)]
pub struct MigrationReport {
    pub state: usize,
    pub project_names: usize,
    pub rss_projects: usize,
    pub licence: usize,
    pub provenance: usize,
    /// Memories given v7 identity fields they lacked.
    pub identity: usize,
    /// Content hashes shared by more than one memory. Expected, not an
    /// error: a local dedup opportunity, and it affects network collapse.
    pub content_hash_collisions: usize,
}

/// key -> fields, for every memory, in key order.
type Records = BTreeMap<String, Fields>;

fn namespace_of(key: &str) -> &str {
    key.split(':').nth(1).unwrap_or("")
}

fn nonempty<'a>(fields: &'a Fields, name: &str) -> Option<&'a str> {
    fields
        .get(name)
        .map(String::as_str)
        .filter(|v| !v.is_empty())
}

/// Sets `name` to `value` on `key`, recording that the record changed.
fn stamp(
    records: &mut Records,
    changed: &mut BTreeMap<String, Fields>,
    key: &str,
    name: &str,
    value: &str,
) {
    if let Some(fields) = records.get_mut(key) {
        fields.insert(name.to_owned(), value.to_owned());
    }
    changed
        .entry(key.to_owned())
        .or_default()
        .insert(name.to_owned(), value.to_owned());
}

pub(crate) fn missing_state(
    records: &mut Records,
    changed: &mut BTreeMap<String, Fields>,
) -> usize {
    let keys: Vec<String> = records
        .iter()
        .filter(|(k, f)| {
            matches!(
                namespace_of(k),
                "episodic" | "project" | "knowledge" | "preference"
            ) && nonempty(f, "state").is_none()
        })
        .map(|(k, _)| k.clone())
        .collect();
    for key in &keys {
        stamp(records, changed, key, "state", "active");
    }
    keys.len()
}

pub(crate) fn project_names(
    records: &mut Records,
    changed: &mut BTreeMap<String, Fields>,
) -> usize {
    let updates: Vec<(String, String)> = records
        .iter()
        .filter(|(k, f)| namespace_of(k) == "project" && nonempty(f, "project_name").is_none())
        .filter_map(|(k, f)| nonempty(f, "project").map(|p| (k.clone(), p.to_owned())))
        .collect();
    for (key, project) in &updates {
        stamp(records, changed, key, "project_name", project);
    }
    updates.len()
}

pub(crate) fn rss_article_projects(
    records: &mut Records,
    changed: &mut BTreeMap<String, Fields>,
) -> usize {
    let keys: Vec<String> = records
        .iter()
        .filter(|(k, f)| {
            namespace_of(k) == "knowledge"
                && nonempty(f, "feed_name").is_some()
                && nonempty(f, "project").is_none()
        })
        .map(|(k, _)| k.clone())
        .collect();
    for key in &keys {
        stamp(records, changed, key, "project", RSS_PROJECT_LABEL);
    }
    keys.len()
}

/// `migrate_licence`: honest defaults. Conversation namespaces are own;
/// articles, imports and untraceable knowledge are unknown; extracted facts
/// inherit their source, resolved from what this pass stamped first and the
/// stored value second, and own when the source is gone.
pub(crate) fn licence(records: &mut Records, changed: &mut BTreeMap<String, Fields>) -> usize {
    let mut own = Vec::new();
    let mut imported = Vec::new();
    let mut pending: Vec<(String, String)> = Vec::new();
    for (key, f) in records.iter() {
        if !matches!(namespace_of(key), "episodic" | "project" | "preference")
            || nonempty(f, "licence").is_some()
        {
            continue;
        }
        if nonempty(f, "imported_at").is_some() {
            imported.push(key.clone());
        } else if let Some(source) = nonempty(f, "enriched_from") {
            pending.push((key.clone(), source.to_owned()));
        } else {
            own.push(key.clone());
        }
    }
    let mut unknown_knowledge = Vec::new();
    for (key, f) in records.iter() {
        if namespace_of(key) != "knowledge" || nonempty(f, "licence").is_some() {
            continue;
        }
        match nonempty(f, "enriched_from") {
            Some(source)
                if nonempty(f, "feed_name").is_none() && nonempty(f, "imported_at").is_none() =>
            {
                pending.push((key.clone(), source.to_owned()));
            }
            _ => unknown_knowledge.push(key.clone()),
        }
    }

    // What a fact's source reads as: this pass's own/imported stamps, else
    // the value already stored. Taken before any fact is assigned, as the
    // Python migration reads the store before writing the facts.
    let mut resolved: HashMap<String, Option<String>> = HashMap::new();
    for key in &own {
        resolved.insert(key.clone(), Some(LICENCE_OWN.to_owned()));
    }
    for key in &imported {
        resolved.insert(key.clone(), Some(LICENCE_UNKNOWN.to_owned()));
    }
    for (_, source) in &pending {
        resolved.entry(source.clone()).or_insert_with(|| {
            records
                .get(source)
                .and_then(|f| nonempty(f, "licence"))
                .map(str::to_owned)
        });
    }

    let mut total = 0;
    for key in own {
        stamp(records, changed, &key, "licence", LICENCE_OWN);
        total += 1;
    }
    for key in imported.into_iter().chain(unknown_knowledge) {
        stamp(records, changed, &key, "licence", LICENCE_UNKNOWN);
        total += 1;
    }
    for (key, source) in pending {
        let value = resolved
            .get(&source)
            .cloned()
            .flatten()
            .unwrap_or_else(|| LICENCE_OWN.to_owned());
        stamp(records, changed, &key, "licence", &value);
        total += 1;
    }
    total
}

fn is_project_context(key: &str, f: &Fields) -> bool {
    if let Some(name) = nonempty(f, "project_name")
        && key == format!("mem:project:{name}")
    {
        return true;
    }
    nonempty(f, "stack").is_some() || nonempty(f, "goals").is_some()
}

/// `migrate_provenance`: articles retrieved, preferences and project context
/// asserted, facts inherit (concluded when the source is gone), everything
/// else, every episodic memory included, concluded.
pub(crate) fn provenance(records: &mut Records, changed: &mut BTreeMap<String, Fields>) -> usize {
    let mut assigned: BTreeMap<String, &'static str> = BTreeMap::new();
    let mut pending: Vec<(String, String)> = Vec::new();
    for ns in ["episodic", "project", "preference", "knowledge"] {
        for (key, f) in records.iter().filter(|(k, _)| namespace_of(k) == ns) {
            if nonempty(f, "provenance").is_some() {
                continue;
            }
            if nonempty(f, "feed_name").is_some() {
                assigned.insert(key.clone(), PROVENANCE_RETRIEVED);
            } else if let Some(source) = nonempty(f, "enriched_from") {
                pending.push((key.clone(), source.to_owned()));
            } else if ns == "preference" || (ns == "project" && is_project_context(key, f)) {
                assigned.insert(key.clone(), PROVENANCE_ASSERTED);
            } else {
                assigned.insert(key.clone(), PROVENANCE_CONCLUDED);
            }
        }
    }
    let mut facts: Vec<(String, String)> = Vec::new();
    for (key, source) in pending {
        let value = assigned
            .get(&source)
            .map(|v| (*v).to_owned())
            .or_else(|| {
                records
                    .get(&source)
                    .and_then(|f| nonempty(f, "provenance"))
                    .map(str::to_owned)
            })
            .unwrap_or_else(|| PROVENANCE_CONCLUDED.to_owned());
        facts.push((key, value));
    }
    let total = assigned.len() + facts.len();
    for (key, value) in assigned {
        stamp(records, changed, &key, "provenance", value);
    }
    for (key, value) in facts {
        stamp(records, changed, &key, "provenance", &value);
    }
    total
}

/// v7 spec §7: content_hash, epoch 1, origin_id, default classification.
pub(crate) fn identity(
    records: &mut Records,
    changed: &mut BTreeMap<String, Fields>,
    origin_id: &str,
) -> (usize, usize) {
    let mut touched = 0;
    let keys: Vec<String> = records.keys().cloned().collect();
    for key in &keys {
        let f = &records[key];
        let mut updates: Vec<(&str, String)> = Vec::new();
        if let Some(content) = nonempty(f, "content")
            && nonempty(f, "content_hash").is_none()
        {
            updates.push(("content_hash", content_hash(content)));
        }
        if nonempty(f, "epoch").is_none() {
            updates.push(("epoch", "1".to_owned()));
        }
        if nonempty(f, "origin_id").is_none() {
            updates.push(("origin_id", origin_id.to_owned()));
        }
        if nonempty(f, "classification").is_none() {
            updates.push(("classification", DEFAULT_CLASSIFICATION.to_owned()));
        }
        if !updates.is_empty() {
            touched += 1;
        }
        for (name, value) in updates {
            stamp(records, changed, key, name, &value);
        }
    }
    let mut hashes: HashMap<&str, usize> = HashMap::new();
    for f in records.values() {
        if let Some(h) = nonempty(f, "content_hash") {
            *hashes.entry(h).or_default() += 1;
        }
    }
    let collisions = hashes.values().filter(|n| **n > 1).count();
    (touched, collisions)
}

impl Store {
    /// Run every migration over the whole store.
    pub fn run_migrations(&self) -> Result<MigrationReport> {
        let mut records: Records = BTreeMap::new();
        {
            let conn = self.conn();
            let mut stmt = conn.prepare("SELECT key, fields FROM memories ORDER BY key")?;
            let mut rows = stmt.query([])?;
            while let Some(row) = rows.next()? {
                let key: String = row.get(0)?;
                let raw: String = row.get(1)?;
                let fields = parse_fields(&key, &raw)?;
                records.insert(key, fields);
            }
        }

        let mut changed: BTreeMap<String, Fields> = BTreeMap::new();
        let mut report = MigrationReport {
            state: missing_state(&mut records, &mut changed),
            project_names: project_names(&mut records, &mut changed),
            rss_projects: rss_article_projects(&mut records, &mut changed),
            ..MigrationReport::default()
        };
        report.licence = licence(&mut records, &mut changed);
        report.provenance = provenance(&mut records, &mut changed);
        (report.identity, report.content_hash_collisions) =
            identity(&mut records, &mut changed, &self.origin_id);

        if !changed.is_empty() {
            let mut conn = self.conn();
            let tx = conn.transaction()?;
            for (key, updates) in &changed {
                merge_memory(&tx, key, memory_namespace(key)?, updates, &self.origin_id)?;
            }
            tx.commit()?;
        }
        if report != MigrationReport::default() {
            info!(?report, "migrations applied");
        }
        Ok(report)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn records(entries: &[(&str, &[(&str, &str)])]) -> Records {
        entries
            .iter()
            .map(|(k, fields)| {
                (
                    (*k).to_owned(),
                    fields
                        .iter()
                        .map(|(a, b)| ((*a).to_owned(), (*b).to_owned()))
                        .collect(),
                )
            })
            .collect()
    }

    fn get<'a>(r: &'a Records, key: &str, field: &str) -> Option<&'a str> {
        r.get(key).and_then(|f| f.get(field)).map(String::as_str)
    }

    #[test]
    fn state_names_and_rss_labels() {
        let mut r = records(&[
            ("mem:episodic:a", &[("content", "x")]),
            ("mem:episodic:b", &[("state", "archived")]),
            ("mem:project:c", &[("project", "omnimem")]),
            ("mem:knowledge:d", &[("feed_name", "n8n Blog")]),
            (
                "mem:knowledge:e",
                &[("feed_name", "Other"), ("project", "Mine")],
            ),
        ]);
        let mut changed = BTreeMap::new();
        assert_eq!(missing_state(&mut r, &mut changed), 4);
        assert_eq!(get(&r, "mem:episodic:b", "state"), Some("archived"));
        assert_eq!(project_names(&mut r, &mut changed), 1);
        assert_eq!(get(&r, "mem:project:c", "project_name"), Some("omnimem"));
        assert_eq!(rss_article_projects(&mut r, &mut changed), 1);
        assert_eq!(get(&r, "mem:knowledge:d", "project"), Some("RSS"));
        assert_eq!(get(&r, "mem:knowledge:e", "project"), Some("Mine"));
    }

    #[test]
    fn licence_is_honest() {
        let mut r = records(&[
            ("mem:episodic:own", &[("content", "a decision")]),
            ("mem:episodic:imp", &[("imported_at", "1")]),
            ("mem:episodic:set", &[("licence", "open")]),
            ("mem:knowledge:article", &[("feed_name", "Feed")]),
            ("mem:knowledge:plain", &[("content", "no origin")]),
            (
                "mem:knowledge:fact-of-own",
                &[("enriched_from", "mem:episodic:own")],
            ),
            (
                "mem:knowledge:fact-of-set",
                &[("enriched_from", "mem:episodic:set")],
            ),
            (
                "mem:knowledge:fact-of-imported",
                &[("enriched_from", "mem:episodic:imp")],
            ),
            (
                "mem:preference:fact-of-gone",
                &[("enriched_from", "mem:episodic:deleted")],
            ),
            (
                "mem:knowledge:imported-fact",
                &[("enriched_from", "mem:episodic:own"), ("imported_at", "1")],
            ),
        ]);
        licence(&mut r, &mut BTreeMap::new());
        let l = |k: &str| get(&r, k, "licence");
        assert_eq!(l("mem:episodic:own"), Some("own"));
        assert_eq!(l("mem:episodic:imp"), Some("unknown"));
        assert_eq!(
            l("mem:episodic:set"),
            Some("open"),
            "a set value is never revisited"
        );
        assert_eq!(l("mem:knowledge:article"), Some("unknown"));
        assert_eq!(l("mem:knowledge:plain"), Some("unknown"));
        assert_eq!(l("mem:knowledge:fact-of-own"), Some("own"));
        assert_eq!(l("mem:knowledge:fact-of-set"), Some("open"));
        assert_eq!(l("mem:knowledge:fact-of-imported"), Some("unknown"));
        assert_eq!(l("mem:preference:fact-of-gone"), Some("own"));
        assert_eq!(l("mem:knowledge:imported-fact"), Some("unknown"));
    }

    #[test]
    fn provenance_classes() {
        let mut r = records(&[
            ("mem:episodic:e", &[("content", "work")]),
            ("mem:preference:p", &[("content", "always")]),
            ("mem:project:omnimem", &[("project_name", "omnimem")]),
            ("mem:project:01ULID", &[("project_name", "omnimem")]),
            ("mem:project:legacy", &[("stack", "python")]),
            ("mem:knowledge:a", &[("feed_name", "Feed")]),
            (
                "mem:knowledge:fact",
                &[("enriched_from", "mem:preference:p")],
            ),
            (
                "mem:knowledge:orphan",
                &[("enriched_from", "mem:episodic:gone")],
            ),
            ("mem:knowledge:kept", &[("provenance", "asserted")]),
        ]);
        provenance(&mut r, &mut BTreeMap::new());
        let p = |k: &str| get(&r, k, "provenance");
        assert_eq!(p("mem:episodic:e"), Some("concluded"));
        assert_eq!(p("mem:preference:p"), Some("asserted"));
        assert_eq!(p("mem:project:omnimem"), Some("asserted"));
        assert_eq!(p("mem:project:01ULID"), Some("concluded"));
        assert_eq!(p("mem:project:legacy"), Some("asserted"));
        assert_eq!(p("mem:knowledge:a"), Some("retrieved"));
        assert_eq!(p("mem:knowledge:fact"), Some("asserted"));
        assert_eq!(p("mem:knowledge:orphan"), Some("concluded"));
        assert_eq!(p("mem:knowledge:kept"), Some("asserted"));
    }

    #[test]
    fn identity_fills_gaps_and_counts_collisions() {
        let mut r = records(&[
            ("mem:episodic:a", &[("content", "same text")]),
            ("mem:episodic:b", &[("content", "same text  ")]),
            ("mem:skill:gen:x-local", &[("name", "x")]),
            (
                "mem:episodic:c",
                &[
                    ("content", "done"),
                    ("content_hash", "sha256:kept"),
                    ("epoch", "3"),
                    ("origin_id", "o"),
                    ("classification", "{}"),
                ],
            ),
        ]);
        let (touched, collisions) = identity(&mut r, &mut BTreeMap::new(), "node");
        assert_eq!(touched, 3);
        assert_eq!(collisions, 1, "trailing whitespace hashes the same");
        assert_eq!(get(&r, "mem:episodic:a", "epoch"), Some("1"));
        assert_eq!(get(&r, "mem:episodic:a", "origin_id"), Some("node"));
        assert_eq!(get(&r, "mem:skill:gen:x-local", "content_hash"), None);
        assert_eq!(get(&r, "mem:episodic:c", "epoch"), Some("3"));
    }

    #[test]
    fn migrations_are_idempotent_on_a_store() {
        let store = Store::open_in_memory().unwrap();
        store
            .set_fields(
                "mem:knowledge:k",
                &Fields::from([("feed_name".to_owned(), "F".to_owned())]),
            )
            .unwrap();
        let first = store.run_migrations().unwrap();
        assert_eq!(first.licence, 1);
        assert_eq!(first.rss_projects, 1);
        let second = store.run_migrations().unwrap();
        assert_eq!(
            second.licence
                + second.provenance
                + second.state
                + second.rss_projects
                + second.identity,
            0
        );
        let f = store.get("mem:knowledge:k").unwrap().unwrap();
        assert_eq!(f["licence"], "unknown");
        assert_eq!(f["provenance"], "retrieved");
    }
}
